### PR TITLE
[Evaluation] [Builtins] Store builtins as 'Int's

### PR DESCRIPTION
**Pull request by effectfully**
Don't look here yet.

### CODE REVIEW

File: plutus-core/plutus-core/src/PlutusCore/Core/Type.hs [hunk 1]
```diff
@@ -67,13 +68,32 @@ data Type tyname uni ann
...
+    pretty = pretty @fun . renum
+
 data Term tyname name uni fun ann
     = Var ann name -- ^ a named variable
     | TyAbs ann tyname (Kind ann) (Term tyname name uni fun ann)
     | LamAbs ann name (Type tyname uni ann) (Term tyname name uni fun ann)
     | Apply ann (Term tyname name uni fun ann) (Term tyname name uni fun ann)
     | Constant ann (Some (ValueOf uni)) -- ^ a constant term
-    | Builtin ann fun
+    | Builtin ann (BuiltinTag fun)
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:96]: UNPACK? here and in UPLC

File: plutus-core/plutus-core/src/PlutusCore/Constant/Meaning.hs [hunk 2]
```diff
@@ -117,18 +117,19 @@ typeOfBuiltinFunction fun = case toBuiltinMeaning @_ @_ @(Term TyName Name _ _ (
...
 -- | Look up the runtime info of a built-in function during evaluation.
 lookupBuiltin
     :: (MonadError (ErrorWithCause err term) m, AsMachineError err fun, Ix fun)
-    => fun -> BuiltinsRuntime fun val -> m (BuiltinRuntime val)
+    => BuiltinTag fun -> BuiltinsRuntime fun val -> m (BuiltinRuntime val)
 -- @Data.Array@ doesn't seem to have a safe version of @(!)@, hence we use a prism.
-lookupBuiltin fun (BuiltinsRuntime env) = case env ^? ix fun of
-    Nothing  -> throwingWithCause _MachineError (UnknownBuiltin fun) Nothing
+lookupBuiltin (BuiltinTag tag) (BuiltinsRuntime env) = case env ^? ix tag of
+    Nothing  -> throwingWithCause _MachineError (UnknownBuiltin undefined) Nothing
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Constant/Meaning.hs:132]: fixme

File: plutus-core/plutus-core/src/PlutusCore/Constant/Meaning.hs [hunk 3]
```diff
@@ -94,7 +94,7 @@ data BuiltinRuntime term =
 
 -- | A 'BuiltinRuntime' for each builtin from a set of builtins.
 newtype BuiltinsRuntime fun term = BuiltinsRuntime
-    { unBuiltinRuntime :: Array fun (BuiltinRuntime term)
+    { unBuiltinRuntime :: Array Int (BuiltinRuntime term)
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Constant/Meaning.hs:97]: There isn't an optimized version of `Array` for `Int`? Maybe it gets specialized anyway.

**Review by michaelpj — COMMENTED** [review]
Not done I know, but here are a couple of things.

File: plutus-core/plutus-core/src/PlutusCore/Constant/Meaning.hs (hunk 3 above)
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Constant/Meaning.hs:97]: Zachary just helped me out in slack and yeah, [there is](https://hackage.haskell.org/package/primitive-0.7.1.0/docs/Data-Primitive-Array.html).

However we're trying to measure two things here:

1. how much of an improvement it is to store builtins as `Int`s
2. how much of an improvement it is to use a "specialized" version of `Array`

and it makes sense to measure them separately. Particularly given the fact that the indexing function in the `primitive` library does not do bounds checking and we do a shit ton of bounds checking, which is something we could drop without changing the arrays library.

File: plutus-core/plutus-core/src/PlutusCore/Core/Type.hs (hunk 1 above)
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:96]: That [happens by default](https://downloads.haskell.org/%7Eghc/latest/docs/html/users_guide/using-optimisation.html#ghc-flag--funbox-small-strict-fields), but ok.

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:96]: But I did forget to add a bang :facepalm:

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:96]: Ah, well, we don't have bangs in TPLC and it's there in UPLC.

Comment (michaelpj) [reply plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:96]: Yeah, but I think it's good to be explicit, and I think you'll get an error if it doesn't fire? So maybe just in UPLC then.

File: plutus-core/plutus-core/src/PlutusCore/Constant/Meaning.hs (hunk 3 above)
Comment (michaelpj) [reply plutus-core/plutus-core/src/PlutusCore/Constant/Meaning.hs:97]: Yep, fine.

**Comment by effectfully** [comment]
Should I get it all to build, 'cause the `benchmark` command does not work otherwise? Or merge `master` in? Or what?

**Comment by michaelpj** [comment]
I think it should work as long as `plutus-benchmark:validation` builds? Maybe try rebasing on master.

**Comment by michaelpj** [comment]
Tried this in another PR and it was worse.

**Pull request closed without merging**